pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -407,9 +407,12 @@ if [[ -z $_NIXOS_REBUILD_REEXEC && -n $canRun && -z $fast ]]; then
             SHOULD_REEXEC=1
         fi
     else
-        runCmd nix "${flakeFlags[@]}" build --out-link "${tmpDir}/nixos-rebuild" "$flake#$flakeAttr.config.system.build.nixos-rebuild" "${extraBuildFlags[@]}" "${lockFlags[@]}"
-        if p=$(readlink -e "${tmpDir}/nixos-rebuild"); then
-            SHOULD_REEXEC=1
+        targetSystem=$(runCmd nix eval --raw "$flake#$flakeAttr.pkgs.system")
+        if [[ $(runCmd nix show-config --json | jq -r '.system.value') == $targetSystem ]]; then
+            runCmd nix "${flakeFlags[@]}" build --out-link "${tmpDir}/nixos-rebuild" "$flake#$flakeAttr.config.system.build.nixos-rebuild" "${extraBuildFlags[@]}" "${lockFlags[@]}"
+            if p=$(readlink -e "${tmpDir}/nixos-rebuild"); then
+                SHOULD_REEXEC=1
+            fi
```

Comment:
I think the solution should work for both Flakes and non-Flakes systems. Now that `nixos-rebuild` has `--file` and `--attr`, I think it is possible that a non-Flakes system also hits the same issue.